### ISSUE TITLE
Broken model path

### ISSUE THREAD

**Issue by b4dboi** [jcrist/hdfscm#2]
Hi,
thank you for your great work on all modules for the deployment of Jupyterhub using YARN on HADOOP. I found a minor issue at the method `to_api_path` in `utils.py` generates bad path, that is subsequently used to compose item's URL in the Jupyter Notebook tree. In my scenario, the fs path URI starts with the protocol name hdfs:// followed by node address. Due to this input format the URL is then generated like: `http://<ip>:<port>/user/xxx/notebooks/hdfs:/nameservice1/user/xxx/notebooks/mynotebook.ipynb` although the proper should be `http://<ip>:<port>/user/xxx/notebooks/mynotebook.ipynb`.

I made a short one-liner workaround by adding the following line at line 30 in `utils.py`:
```fs_path = fs_path[7+fs_path[7:].find('/'):] if fs_path.startswith('hdfs://') else fs_path```

If you find it useful please add it to the code.

**Comment by jcrist** [comment]
Thanks for filing an issue. Can you provide a bit more information about where this occurs? The protocol and address from the files shouldn't be present in that call. I suspect that you've set `root_dir` or `root_dir_template` with `hdfs://<node-address>/user/xxx/` instead of using just `/user/xxx/`. I could add a check for this in the configuration to provide a better error, but you shouldn't have the protocol or address in these configuration fields - the `hdfs_address` and `hdfs_port` fields are there for this purpose.

**Comment by b4dboi** [comment]
I made no changes to `root_dir` nor `root_dir_template`, leaving them in their default state. I am simply passing this config to the YARN container `c.YarnSpawner.args = ['--NotebookApp.contents_manager_class="hdfscm.HDFSContentsManager"']` 

I run Cloudera cluster and I am using default config to access the HDFS generated by the Cloudera. I don't believe there is some unusual notation in the config itself as everything else in hdfscm module works smoothly.

**Comment by jcrist** [comment]
Hmmm, ok. Can you describe what operation you're doing fails? What triggers the failure? Can you post a traceback as well?

Also, what version of `pyarrow` are you using?

```python
import pyarrow
pyarrow.__version__
```

**Comment by b4dboi** [comment]
**What operation fails?**
As a user of Jupyter notebook, I am not able to open neither notebook nor directory from the tree on Files panel since the generated link contains full hdfs path including protocol, e.g. `http://<ip>:<port>/user/xxx/notebooks/hdfs:/nameservice1/user/xxx/notebooks/mynotebook.ipynb`. Due to the broken link, the user lands on page saying: _404 : Not Found, You are requesting a page that does not exist!_ After changing the link to the proper one (`http://<ip>:<port>/user/xxx/notebooks/mynotebook.ipynb`) there is no problem with opening the notebook or showing content of directory in the tree of files.

**What triggers the failure?**
User's click on any file or directory link in the Files panel of Jupyter notebook.

**Traceback**
```
hdfsExists: invokeMethod((Lorg/apache/hadoop/fs/Path;)Z) error:
IllegalArgumentException: Pathname /user/xxx/notebooks/hdfs:/nameservice1/user/xxx/notebooks/notebook.ipynb from /user/xxx/notebooks/hdfs:/nameservice1/user/xxx/notebooks/notebook.ipynb is not a valid DFS filename.java.lang.IllegalArgumentException: Pathname /user/xxx/notebooks/hdfs:/nameservice1/user/xxx/notebooks/notebook.ipynb from /user/xxx/notebooks/hdfs:/nameservice1/user/xxx/notebooks/notebook.ipynb is not a valid DFS filename.
	at org.apache.hadoop.hdfs.DistributedFileSystem.getPathName(DistributedFileSystem.java:233)
	at org.apache.hadoop.hdfs.DistributedFileSystem$29.doCall(DistributedFileSystem.java:1495)
	at org.apache.hadoop.hdfs.DistributedFileSystem$29.doCall(DistributedFileSystem.java:1492)
	at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81)
	at org.apache.hadoop.hdfs.DistributedFileSystem.getFileStatus(DistributedFileSystem.java:1507)
	at org.apache.hadoop.fs.FileSystem.exists(FileSystem.java:1668)
[W 2019-04-10 11:08:09.571 YarnSingleUserNotebookApp web:1782] 404 GET /user/xxx/notebooks/hdfs:/nameservice1/user/xxx/notebooks/notebook.ipynb (::ffff:ip): No such file or directory: hdfs:/nameservice1/user/xxx/notebooks/notebook.ipynb
[W 2019-04-10 11:08:09.572 YarnSingleUserNotebookApp log:158] 404 GET /user/xxx/notebooks/hdfs:/nameservice1/user/xxx/notebooks/notebook.ipynb (xxx@::ip) 26.51ms
```

**Versions**
- Cloudera 6.1.0 cluster
- notebook==5.7.8
- jupyterhub==0.9.6
- jupyterhub-yarnspawner==0.2.0
- pyarrow==0.13.0
- skein==0.6.1
- jupyter-hdfscm==0.1.1

**Workaround**
Adding the following line at line 30 in `utils.py`:
```fs_path = fs_path[7+fs_path[7:].find('/'):] if fs_path.startswith('hdfs://') else fs_path```

**Comment by jcrist** [comment]
Odd, I am unable to reproduce the issue with the same versions. I suspect this has nothing to do with the python libraries, and something to do with your version of `libhdfs` (which arrow wraps). Can you run the following and report back? What version of hadoop are you using? I suspect that the `hdfs://` prefix is returned somewhere on your version where it isn't on the one we test with (2.6).

```python
from pyarrow import hdfs

fs = hdfs.connect()

# Some path to a directory containing some files
DIR_TO_CHECK = ...

print("Getting records from ls")
print("Everything below should report `False`")
for record in fs.ls(DIR_TO_CHECK, True):
    print(record['name'].startswith('hdfs'))

print("Getting records from info")
print("Everything below here should report `True`")
for path in fs.ls(DIR_TO_CHECK):
    record = fs.info(path)
    print(record['path'].startswith('hdfs'))
```

**Comment by jcrist** [comment]
@b4dboi any updates here? I now have tested on a CDH 6.2 cluster (as well as a CDH 5.14 cluster) and things work fine on both. I'm not sure why you're experiencing issues while I'm not.

If we can't find a reproducible way to hit this bug, then I'm happy to add a fix around it. The location of your proposed fix isn't where we should handle this, rather things should be normalized sooner in `hdfscm.HDFSContentsManager._model_from_info`. I suspect (but can't be sure) that the traceback you posted above is part of a larger Python traceback - having the full traceback would better help diagnose this issue.

**Comment by b4dboi** [comment]
Sorry, I am on vacation for a few weeks and cannot access my testbed remotely. I will report on that when I get back to the office. Thank you for your patience.

**Comment by b4dboi** [comment]
Unfortunately, my testbed was deleted during my vacation and I have only limited access to production cluster. However, I ran the piece of the pyarrow code you posted above and I received True/False reports as expected (no _hdfs_ included in record['name'], and _hdfs_ included in record['path']). The Hadoop version is Hadoop 3.0.0-cdh6.1.0. 

I'm not able to dig deeper into the plugin right now, as I have no access to cluster running Jupyter Hub for now (it works well in isolated environment with the proposed workaround, though it is not optimal). My options to reproduce this potential bug are very limited. If you consider this issue irelevant anymore feel free to close it.

**Comment by georgepachitariu** [comment]
Hi @jcrist, I found the same bug as mentioned in this thread. The problem arises from the fact that both `ls` and `info` include `hdfs://nameservice1`. Look at this python code:
```
from pyarrow import hdfs
fs = hdfs.connect()
print(fs.ls("/user/gpachitariu/wf_launchJupyterNotebook", False)[0])
print(fs.ls("/user/gpachitariu/wf_launchJupyterNotebook", True)[0])
print(fs.info("/user/gpachitariu/wf_launchJupyterNotebook"))
```
The output is:
```
hdfs://nameservice1/user/gpachitariu/wf_launchJupyterNotebook/activate_env.sh
{'kind': 'file', 'name': 'hdfs://nameservice1/user/gpachitariu/wf_launchJupyterNotebook/activate_env.sh', 'owner': 'gpachitariu', 'group': 'gpachitariu', 'last_modified_time': 1563834916, 'last_access_time': 1564061712, 'size': 1086, 'replication': 3, 'block_size': 134217728, 'permissions': 420}
{'path': 'hdfs://nameservice1/user/gpachitariu/wf_launchJupyterNotebook', 'owner': 'gpachitariu', 'group': 'gpachitariu', 'size': 0, 'block_size': 0, 'last_modified': 1564062463, 'last_accessed': 0, 'replication': 0, 'permissions': 493, 'kind': 'directory'}
```

Please look at the attached file where I fix it. I tested that the change, adding `urlsplit`, works both when `hdfs://nameservice1` exists and when it does not.
[0001-Fixed-path-patch.txt](https://github.com/jcrist/hdfscm/files/3431966/0001-Fixed-path-patch.txt)

My setup: Cloudera 5.14.4 and libhdfs is from here:
http://archive.cloudera.com/cdh5/debian/jessie/amd64/cdh/pool/contrib/h/hadoop/libhdfs0_2.6.0+cdh5.14.4+2785-1.cdh5.14.4.p0.4~jessie-cdh5.14.4_amd64.deb

**Comment by jcrist** [comment]
Hmmm, interesting. It's not clear to me why some environments have the `hdfs://` prefix for `ls` and others don't. The patch looks fine either way, would you be able to submit a PR?

**Comment by georgepachitariu** [comment]
It's not clear no me either. I tried to push my branch to your repo, but I get:
```remote: Permission to jcrist/hdfscm.git denied to georgepachitariu.```

**Comment by jcrist** [comment]
Yes, you need to fork the repo, push to your fork, then make a pull request. If you need more information, here's a walkthrough of how this works: https://www.asmeurer.com/git-workflow/

**Comment by jcrist** [comment]
Fixed by #3.

**Comment by jcrist** [comment]
Version 0.2.0 has been released to PyPI with this fix, and should be on conda-forge shortly.
